models/eprop_readout.cpp

```diff
@@ -0,0 +1,388 @@
+
+void
+eprop_readout::compute_error_signal_mean_squared_error( const long lag )
+{
+  S_.readout_signal_ = S_.v_m_ + P_.E_L_;
+  S_.error_signal_ = S_.readout_signal_ - S_.target_signal_;
+}
+
+/* ----------------------------------------------------------------
+ * Event handling functions
+ * ---------------------------------------------------------------- */
+
+void
+eprop_readout::handle( DelayedRateConnectionEvent& e )
+{
+  const size_t rport = e.get_rport();
+  assert( rport < SUP_RATE_RECEPTOR );
+
+  auto it = e.begin();
+  assert( it != e.end() );
+
+  const double signal = e.get_weight() * e.get_coeffvalue( it );
+  if ( rport == LEARNING_WINDOW_SIG )
+  {
+    S_.learning_window_signal_ = signal;
+  }
+  else if ( rport == TARGET_SIG )
+  {
+    S_.target_signal_ = signal;
+  }
+
+  assert( it == e.end() );
+}
+
+void
+eprop_readout::handle( SpikeEvent& e )
+{
+  assert( e.get_delay_steps() > 0 );
+
+  B_.spikes_.add_value(
+    e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ), e.get_weight() * e.get_multiplicity() );
+}
+
+void
+eprop_readout::handle( CurrentEvent& e )
+{
+  assert( e.get_delay_steps() > 0 );
+
+  B_.currents_.add_value(
+    e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() ), e.get_weight() * e.get_current() );
+}
+
+void
+eprop_readout::handle( DataLoggingRequest& e )
+{
+  B_.logger_.handle( e );
+}
+
+void
+eprop_readout::compute_gradient( const long t_spike,
+  const long t_spike_previous,
+  double& z_previous_buffer,
+  double& z_bar,
+  double& e_bar,
+  double& epsilon,
+  double& weight,
+  const CommonSynapseProperties& cp,
+  WeightOptimizer* optimizer )
+{
+  double z = 0.0;                // spiking variable
+  double z_current_buffer = 1.0; // buffer containing the spike that triggered the current integration
+  double L = 0.0;                // error signal
+  double grad = 0.0;             // gradient
+
+  const EpropSynapseCommonProperties& ecp = static_cast< const EpropSynapseCommonProperties& >( cp );
+  const auto optimize_each_step = ( *ecp.optimizer_cp_ ).optimize_each_step_;
+
+  auto eprop_hist_it = get_eprop_history( t_spike_previous - 1 );
+
+  const long t_compute_until = std::min( t_spike_previous + P_.eprop_isi_trace_cutoff_, t_spike );
+
+  for ( long t = t_spike_previous; t < t_compute_until; ++t, ++eprop_hist_it )
+  {
+    z = z_previous_buffer;
+    z_previous_buffer = z_current_buffer;
+    z_current_buffer = 0.0;
+
+    L = eprop_hist_it->error_signal_;
+
+    z_bar = V_.P_v_m_ * z_bar + V_.P_z_in_ * z;
+
+    if ( optimize_each_step )
+    {
+      grad = L * z_bar;
+      weight = optimizer->optimized_weight( *ecp.optimizer_cp_, t, grad, weight );
+    }
+    else
+    {
+      grad += L * z_bar;
+    }
+  }
+
+  if ( not optimize_each_step )
+  {
+    weight = optimizer->optimized_weight( *ecp.optimizer_cp_, t_compute_until, grad, weight );
+  }
+
+  const int power = t_spike - ( t_spike_previous + P_.eprop_isi_trace_cutoff_ );
```

Comment:
For consistency with all other variables:
```suggestion
  const long power = t_spike - ( t_spike_previous + P_.eprop_isi_trace_cutoff_ );
```

Reply:
Fixed in [00c80ee](https://github.com/nest/nest-simulator/pull/3207/commits/00c80ee25dedcd375737dbd2fb0649d05751aac7)

Reply:
Fixed in [00c80ee](https://github.com/nest/nest-simulator/pull/3207/commits/00c80ee25dedcd375737dbd2fb0649d05751aac7)